src/lib/registration/controllers/package_search.rb

```diff
@@ -0,0 +1,180 @@
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "registration/package_search"
+
+Yast.import "Popup"
+Yast.import "HTML"
+
+module Registration
+  module Controllers
+    # Implements the actions and keeps the state for the package search feature
+    class PackageSearch
+      include Yast::I18n
+
+      # @return [Array<RemotePackage>] List of selected packages
+      attr_reader :selected_packages
+
+      # Constructor
+      def initialize
+        textdomain "registration"
+        @selected_packages = []
+      end
+
+      # Performs a package search
+      #
+      # @param text [String] Term to search for
+      # @return [Array<Registration::RemotePackage>] List of packages
+      def search(text)
+        @search = ::Registration::PackageSearch.new(text: text)
+        @search.packages
+      end
+
+      # Selects/unselects the current package for installation
+      #
+      # It does nothing if the package is already installed.
+      def toggle_package(package)
+        return if package.installed?
+
+        if package.selected?
+          unselect_package(package)
+        else
+          select_package(package)
+        end
+      end
+
+    private
+
+      # Selects the current package for installation
+      #
+      # If required, it selects the corresponding addon
+      #
+      # @param package [RemotePackage] Package to select
+      def select_package(package)
+        addon = package.addon
+        select_addon(addon) if addon
+        set_package_as_selected(package) if addon.nil? || addon.selected? || addon.registered?
+      end
+
+      # Unselects the current package for installation
+      #
+      # If not needed, unselects the corresponding addon
+      #
+      # @param package [RemotePackage] Package to unselect
+      #
+      # @see #unselect_addon
+      # @see #unselect_package!
+      def unselect_package(package)
+        unset_package_as_selected(package)
+        unselect_addon(package.addon) if package.addon
+      end
+
+      # Selects the given addon if needed
+      #
+      # @param addon [Addon] Addon to select
+      def select_addon(addon)
+        return if addon.registered? || addon.selected? || addon.auto_selected?
+        addon.selected if enable_addon?(addon)
+      end
+
+      # Unselects the given addon if required
+      #
+      # @param addon [Addon] Addon to unselect
+      def unselect_addon(addon)
+        return if addon.registered? || needed_addon?(addon)
+        addon.unselected if disable_addon?(addon)
```

Comment:
in general as I see there a lot of conditions, I miss a bit more logging there to see what happen and why. Especially if issue appears on customer machine.